### ISSUE TITLE
TrilinosCouplings: fenl_ensemble.hpp errors

### ISSUE THREAD

**Issue by sagitter** [trilinos/Trilinos#837]
Hi all.

`fenl` example does not compile with following error:

```
In file included from /builddir/build/BUILD/trilinos-12.8.1/Trilinos-trilinos-release-12-8-1/packages/trilinoscouplings/examples/fenl/main_ensemble.cpp:9:0:
/builddir/build/BUILD/trilinos-12.8.1/Trilinos-trilinos-release-12-8-1/packages/trilinoscouplings/examples/fenl/fenl_ensemble.hpp:164:9: error: 'GetMeanValsFunc' is not a class template
   class GetMeanValsFunc< Kokkos::View< Sacado::MP::Vector<Storage>*,
         ^
/builddir/build/BUILD/trilinos-12.8.1/Trilinos-trilinos-release-12-8-1/packages/trilinoscouplings/examples/fenl/fenl_ensemble.hpp:248:15: error: 'SolverManager' in namespace 'Belos' does not name a type
   apply(const Belos::SolverManager<Sc,V,O>& solver) {
               ^
/builddir/build/BUILD/trilinos-12.8.1/Trilinos-trilinos-release-12-8-1/packages/trilinoscouplings/examples/fenl/fenl_ensemble.hpp:248:22: error: ISO C++ forbids declaration of 'parameter' with no type [-fpermissive]
   apply(const Belos::SolverManager<Sc,V,O>& solver) {
                      ^
/builddir/build/BUILD/trilinos-12.8.1/Trilinos-trilinos-release-12-8-1/packages/trilinoscouplings/examples/fenl/fenl_ensemble.hpp:248:35: error: expected ',' or '...' before '<' token
   apply(const Belos::SolverManager<Sc,V,O>& solver) {
                                   ^
/builddir/build/BUILD/trilinos-12.8.1/Trilinos-trilinos-release-12-8-1/packages/trilinoscouplings/examples/fenl/fenl_ensemble.hpp: In static member function 'static std::vector<int, std::allocator<int> > Kokkos::Example::ExtractEnsembleIts<Sacado::MP::Vector<S>, V, O>::apply(int)':
/builddir/build/BUILD/trilinos-12.8.1/Trilinos-trilinos-release-12-8-1/packages/trilinoscouplings/examples/fenl/fenl_ensemble.hpp:249:11: error: 'PseudoBlockCGSolMgr' in namespace 'Belos' does not name a type
     const Belos::PseudoBlockCGSolMgr<Sc, V, O>* cg_solver =
           ^
/builddir/build/BUILD/trilinos-12.8.1/Trilinos-trilinos-release-12-8-1/packages/trilinoscouplings/examples/fenl/fenl_ensemble.hpp:251:9: error: 'cg_solver' was not declared in this scope
     if (cg_solver != 0)
         ^
/usr/bin/cmake -E cmake_progress_report /builddir/build/BUILD/trilinos-12.8.1/Trilinos-trilinos-release-12-8-1/build/CMakeFiles 
```

`RHEL7 PPC64-LittleEndian` buildroot, `GCC 4.8.5`
`cmake` configuration:

`+ /usr/bin/cmake -DCMAKE_C_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_Fortran_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_INSTALL_PREFIX:PATH=/usr -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib64 -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DLIB_SUFFIX=64 -DBUILD_SHARED_LIBS:BOOL=ON -DCMAKE_BUILD_TYPE:STRING=Release '-DCMAKE_C_FLAGS_RELEASE:STRING=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mcpu=power8 -mtune=power8 -Wl,-z,relro -fPIC -Wl,-z,now' '-DCMAKE_CXX_FLAGS_RELEASE:STRING=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mcpu=power8 -mtune=power8 -Wl,-z,relro -fPIC -Wl,-z,now' '-DCMAKE_Fortran_FLAGS_RELEASE:STRING=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mcpu=power8 -mtune=power8 -Wl,-z,relro -fPIC -Wl,-z,now' '-DCMAKE_EXE_LINKER_FLAGS:STRING=-Wl,-z,relro  -fPIC -pie -Wl,-z,now -Wl,--as-needed' -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_SKIP_RPATH:BOOL=YES -DCMAKE_SKIP_INSTALL_RPATH:BOOL=YES -DBUILD_SHARED_LIBS:BOOL=ON -DTPL_ENABLE_MPI=OFF -DTrilinos_ENABLE_FORTRAN:BOOL=ON -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_ENABLE_OpenMP:BOOL=ON -DTrilinos_ENABLE_PyTrilinos:BOOL=OFF -DTpetra_INST_SERIAL:BOOL=ON -DCMAKE_VERBOSE_MAKEFILE=TRUE -DTrilinos_VERBOSE_CONFIGURE=OFF -DTrilinos_ENABLE_CXX11=ON -DTrilinos_ENABLE_MueLu:BOOL=OFF -DTPL_Boost_INCLUDE_DIRS:PATH=/usr/include/boost148 -DTPL_Boost_LIBRARIES:PATH=/usr/lib64/boost148/libboost_system.so:/usr/lib64/boost148/libboost_program_options.so:/usr/lib64/boost148/libboost_signals.so:/usr/lib64/boost148/libboost_regex.so:/usr/lib64/boost148/libboost_filesystem.so:/usr/lib64/boost148/libboost_thread.so:/usr/lib64/boost148/libboost_wserialization.so:/usr/lib64/boost148/libboost_serialization.so -DTrilinos_INSTALL_INCLUDE_DIR=/usr/include/trilinos -DTrilinos_INSTALL_LIB_DIR=/usr/lib64 -DTrilinos_INSTALL_RUNTIME_DIR=/usr/bin -DTrilinos_INSTALL_EXAMPLE_DIR=/usr/share/trilinos/examples -DTrilinos_ENABLE_STRONG_C_COMPILE_WARNINGS=OFF -DTrilinos_ENABLE_STRONG_CXX_COMPILE_WARNINGS=OFF -DTrilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON -DTPL_ENABLE_gtest:BOOL=OFF -DIsorropia_ENABLE_EXAMPLES:BOOL=OFF -DTrilinos_ENABLE_STK:BOOL=OFF -DTrilinos_ENABLE_ALL_PACKAGES=ON ..`

**Comment by mhoemmen** [comment]
I'm looking at the following build log:

https://kojipkgs.fedoraproject.org//work/tasks/2528/16402528/build.log

If that still applies, it looks like Belos, Kokkos, and Stokhos are all enabled.  If so, then this is a build error.  Are those packages all enabled in your build?

**Comment by mhoemmen** [comment]
btw, you're probably wondering why Trilinos has build errors like this ;-)  Trilinos is more like a Linux distribution than a software library.  It has a lot of packages.  We test a subset of packages nightly, but most users don't enable or install all the packages.  We're currently expanding the set of packages that we build and test by default.

**Comment by mhoemmen** [comment]
@trilinos/stokhos @trilinos/belos

**Comment by sagitter** [comment]
>I'm looking at the following build log:
>
>https://kojipkgs.fedoraproject.org//work/tasks/2528/16402528/build.log

No, it's a new build:
https://kojipkgs.fedoraproject.org//work/tasks/6500/16436500/build.log

All packages enabled except `MueLu` and `STK`.

**Comment by mhoemmen** [comment]
`Final set of enabled packages:  Gtest ThreadPool Kokkos Teuchos RTOp Sacado Epetra Zoltan Shards GlobiPack Triutils Tpetra EpetraExt Thyra Xpetra OptiPack Isorropia Pliris AztecOO Galeri Amesos Pamgen Zoltan2 Ifpack ML Belos ShyLU Amesos2 SEACAS Komplex Anasazi Ifpack2 Stratimikos FEI Teko Intrepid Intrepid2 Phalanx NOX Moertel Rythmos Stokhos ROL Piro Panzer TrilinosCouplings Pike 47`

I didn't realize one could build Stokhos with MueLu disabled.  That's a new one for me ;-)  (Tpetra is my package, so I test all of its downstream dependencies, including MueLu and Stokhos.)

Thanks for posting the updated build log!  We'll check on this one.

@etphipp @trilinos/sacado @trilinos/stokhos @trilinos/belos

**Comment by etphipp** [comment]
I looked in the fenl example and the relevant includes are protected by this:

#if defined(HAVE_TRILINOSCOUPLINGS_BELOS) && defined(HAVE_TRILINOSCOUPLINGS_MUELU)

so I need to protect the corresponding specializations in fenl_ensemble.hpp.  I will fix that.

What branch are you working from?  Does this need to go into the latest release branch?

**Comment by mhoemmen** [comment]
@etphipp wrote:
> What branch are you working from? Does this need to go into the latest release branch?

I think @sagitter is working on a Fedora "package" for Trilinos.  It's likely that they will want to work from a release tarball.

**Comment by sagitter** [comment]
>>What branch are you working from? Does this need to go into the latest release branch?
>
>I think @sagitter is working on a Fedora "package" for Trilinos. It's likely that they will want to work from >a release tarball.

I'm working on Fedora and RHEL7.
This specific issue is related to a [RHEL7](https://fedoraproject.org/wiki/EPEL) build on `PPC64le` architecture but i think it happens on any architectures.

All my tests are done with [Trilinos-12.8.1](https://github.com/trilinos/Trilinos/releases/tag/trilinos-release-12-8-1).

**Comment by etphipp** [comment]
This is now fixed on develop.

**Comment by mhoemmen** [comment]
@sagitter , I think I've been making the mistake of closing issues that need back-porting to the 12.8 release branch...

**Comment by sagitter** [comment]
I'm finishing `Trilinos` package, before begin the review on Fedora i think `Trilinos-12.10` will be ready.

**Comment by mhoemmen** [comment]
@sagitter wrote:
> I'm finishing Trilinos package, before begin the review on Fedora i think Trilinos-12.10 will be ready.

If you don't mind using 12.10 :-)  We'll still have to pack-port to the 12.10 release branch, though, since we've already tagged the branch.

**Comment by etphipp** [comment]
So is applying this to the 12.10 branch sufficient?

**Comment by sagitter** [comment]
I am disabling `TrilinosCouplings` test in `Trilinos-12.8.1`; it's not a problem for me.

**Comment by etphipp** [comment]
I just merged this into the 12.10 branch as well, so you should be good to go when you get to working from that release.

I am closing this for now.  If this needs to be moved to another branch as well.  Please re-open it.
